src/CachePeer.cc

```diff
@@ -114,6 +117,208 @@ CachePeer::connectTimeout() const
     return Config.Timeout.peer_connect;
 }
 
+void
+CachePeer::dumpOptions(std::ostream &os) const
+{
+    if (options.proxy_only)
+        os << " proxy-only";
+
+    if (options.no_query)
+        os << " no-query";
+
+    if (options.background_ping)
+        os << " background-ping";
+
+    if (options.no_digest)
+        os << " no-digest";
+
+    if (options.default_parent)
+        os << " default";
+
+    if (options.roundrobin)
+        os << " round-robin";
+
+    if (options.carp)
+        os << " carp";
+
+#if USE_AUTH
+    if (options.userhash)
+        os << " userhash";
+#endif
+
+    if (options.sourcehash)
+        os << " sourcehash";
+
+    if (options.weighted_roundrobin)
+        os << " weighted-round-robin";
+
+    if (options.mcast_responder)
+        os << " multicast-responder";
+
+#if PEER_MULTICAST_SIBLINGS
+    if (options.mcast_siblings)
+        os << " multicast-siblings";
+#endif
+
+    if (weight != 1)
+        os << " weight=" << weight;
+
+    if (options.closest_only)
+        os << " closest-only";
+
+#if USE_HTCP
+    if (options.htcp)
+    {
+        std::vector<const char *> opts {};
+        if (options.htcp_oldsquid)
+            opts.push_back("oldsquid");
+        if (options.htcp_no_clr)
+            opts.push_back("no-clr");
+        if (options.htcp_no_purge_clr)
+            opts.push_back("no-purge-clr");
+        if (options.htcp_only_clr)
+            opts.push_back("only-clr");
+        os << AsList(opts).prefixedBy(" htcp=").delimitedBy(",");
+    }
+#endif
+
+    if (options.no_netdb_exchange)
+        os << " no-netdb-exchange";
+
+#if USE_DELAY_POOLS
+    if (options.no_delay)
+        os << " no-delay";
+#endif
+
+    if (login)
+        os << " login=" << login;
+
+    if (mcast.ttl > 0)
+        os << " ttl=" << mcast.ttl;
+
+    if (connect_timeout_raw > 0)
+        os << " connect-timeout=" << connect_timeout_raw;
+
+    if (connect_fail_limit != PEER_TCP_MAGIC_COUNT)
+        os << " connect-fail-limit=" << connect_fail_limit;
+
+#if USE_CACHE_DIGESTS
+    if (digest_url)
+        os << " digest-url=" << digest_url;
+#endif
+
+    if (options.allow_miss)
+        os << " allow-miss";
+
+    if (options.no_tproxy)
+        os << " no-tproxy";
+
+    if (max_conn > 0)
+        os << " max-conn=" << max_conn;
+
+    if (standby.limit > 0)
+        os << " standby=" << standby.limit;
+
+    if (options.originserver)
+        os << " originserver";
+
+    if (domain)
+        os << " forceddomain=" << domain;
+
+    if (connection_auth == 0)
+        os << " connection-auth=off";
+    else if (connection_auth == 1)
+        os << " connection-auth=on";
+    else if (connection_auth == 2)
+        os << " connection-auth=auto";
+
+    secure.dumpCfg(os, "tls-");
+}
+
+void
+CachePeer::reportStatistics (std::ostream& os)
+{
+    os <<
+       "- " << name << ":\n" <<
```

Comment:
```
- localhost:
  type: sibling

- localhost.localdomain:
  type: sibling
```

> it's valid syntax.

Thank you for doing the legwork required to reach this important milestone!

The current valid YAML output is also more-or-less human-friendly -- another requirement met.

Unfortunately, the current output is not machine-friendly enough. With some YAML/JSON libraries, this output is nearly unusable because getting the name of a cache_peer would require such heroic (and fragile) efforts as enumerating all cache_peer hash keys to find the key that is actually a cache_peer name.

Fortunately, this is easy to fix. I suggest the following as a starting point:


```suggestion
    os << "-";
    
    os << "  name: " << name << "\n" <<
```

or, to make each cache_peer reporting line format the same (at the expense of more lines in the report), add a new line after "-"

```suggestion
    os << "-\n";
    
    os << "  name: " << name << "\n" <<
```

Reply:
It's even easier than that. Since the peer name is unique due to how config works, we can just do away with the array and have a map, keyed on the peer name.

This is what it looks like:
```yaml
localhost:
  type: sibling
  host: localhost/3129/0
  flags: default connection-auth=auto tls-disable
  addresses: [ "[::1]" ]
  status: Down
  fetches: 0
  open connections: 0
  average RTT: 0 msec
  pings sent: 0
  pings acked: 0 0%
  replies ignored: 0 0%
  last failed connection at: 10/Apr/2024:18:00:55 +0700
  keep-alive ratio: 0%

localhost.localdomain:
  type: sibling
  host: localhost.localdomain/3130/0
  flags: default connection-auth=auto tls-disable
  addresses: [  ]
  status: Down
  fetches: 0
  open connections: 0
  average RTT: 0 msec
  pings sent: 0
  pings acked: 0 0%
  replies ignored: 0 0%
  keep-alive ratio: 0%
```
It is valid yaml, and should address your concern

Reply:
> It is valid yaml, and should address your concern

Agreed, but there is a catch. If we switch to a map, we are making the task of restoring configured cache_dir order (which is important in some contexts) much more difficult for consumers that load this report into an order-destroying hash (because that is what their yaml2foo library provides for YAML maps). If you switch to a map, please add a new field to each cache_dir entry to specify a 1-based configuration index (e.g., "configuration index: 2" for localhost.localdomain cache_peer in your example).

I am not an YAML conversion expert, so if an expert tells us that my concern is unfounded because all popular converters preserve order in maps, then we do not need to add a configuration index.

N.B. I am aware that we report cache_dirs in configuration order.